Fix recovering metadata from docstore

HollowMan6 opened the pull request:
# Description

Fix recovering metadata from docstore

## Type of Change

- [X] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

Test code:
```bash


from llama_index import VectorStoreIndex, SimpleDirectoryReader
from llama_index.vector_stores import MilvusVectorStore
from llama_index.storage.storage_context import StorageContext
from llama_index.storage.index_store import MongoIndexStore
from llama_index.storage.docstore import MongoDocumentStore

MONGO_URL = "mongodb://localhost:27017"

def metadata(path):
    return {"path": path}

documents = SimpleDirectoryReader('data/paul_graham/', file_metadata=metadata).load_data()
docstore = MongoDocumentStore.from_uri(MONGO_URL)
docstore.add_documents(documents)
index_store = MongoIndexStore.from_uri(MONGO_URL)
vector_store = MilvusVectorStore(overwrite=True)
storage_context = StorageContext.from_defaults(vector_store=vector_store, index_store=index_store, docstore=docstore)
index = VectorStoreIndex.from_documents(documents, storage_context=storage_context)

query_engine = index.as_query_engine()
response = query_engine.query("What did the author learn?")
for n in response.source_nodes:
    print(n.node.metadata)
```

Should display:
```bash
{'path': 'data/paul_graham/paul_graham_essay.txt'}
```

- [ ] Added new unit/integration tests
- [ ] Added new notebook (that tests end-to-end)
- [X] I stared at the code and made sure it makes sense

# Suggested Checklist:

- [X] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

HollowMan6 commented:
Looks like it's not a proper fix, will try to investigate a way.

logan-markewich commented:
@HollowMan6 Actually, I think this issue can be fixed without this PR

Can you try initializing the vector store like this?

```python
index = VectorStoreIndex.from_documents(documents, storage_context=storage_context, store_nodes_override=True)
```

HollowMan6 commented:
Now I switched to Redis, the response does get printed this time, so I think the `None` Response bug for `store_nodes_override=True` is specific to milvus. However, `store_nodes_override=True` still does nothing for me to get the metadata printed correctly, only after applying my patch in this PR can I fix this issue.

Logs:
```bash
INFO:llama_index.vector_stores.redis:Deleting index llama_index
INFO:llama_index.vector_stores.redis:Creating index llama_index
INFO:llama_index.vector_stores.redis:Added 20 documents to index llama_index
INFO:llama_index.token_counter.token_counter:> [build_index_from_nodes] Total LLM token usage: 0 tokens
INFO:llama_index.token_counter.token_counter:> [build_index_from_nodes] Total embedding token usage: 17502 tokens
INFO:llama_index.vector_stores.redis:Using filters: *
INFO:llama_index.vector_stores.redis:Querying index llama_index
INFO:llama_index.vector_stores.redis:Found 2 results for query with id ['llama_index_2b44ed30-2912-4e1a-ad54-f59d2d2ff774', 'llama_index_c1dfe331-6a4a-4b31-a36a-30b8d0bebe4b']
INFO:llama_index.token_counter.token_counter:> [retrieve] Total LLM token usage: 0 tokens
INFO:llama_index.token_counter.token_counter:> [retrieve] Total embedding token usage: 6 tokens
INFO:llama_index.token_counter.token_counter:> [get_response] Total LLM token usage: 2012 tokens
INFO:llama_index.token_counter.token_counter:> [get_response] Total embedding token usage: 0 tokens

The author learned that it is better for technology companies to be run by product people than sales people, that editing code by too many people leads to bugs, that cheap office space is not a bargain if it is depressing, that planned meetings are inferior to corridor conversations, that big, bureaucratic customers are a dangerous source of money, that there is not much overlap between conventional office hours and the optimal time for hacking, that prestige is a danger sign, that the low end eats the high end, that HTML was a derivative of SGML, that art school does not bear the same relationship to art that medical school bears to medicine, that signature styles are the visual equivalent of show business schticks, that painting students are supposed to express themselves, that New York is where the artists are, that painting is post-rigorous, that painting students are confused and demoralized by what they find at RISD, that painting can be done for free, that applications of continuations was a suitable topic for a dissertation, that the Accademia di Belli Arti in Florence was a way of excluding stranieri, that the students and faculty at the Accademia had an arrangement whereby the students wouldn't require the faculty to teach anything, and that the model the author painted also made a living from making fakes for a local antique dealer. <class 'llama_index.response.schema.Response'>
{}
{}
```

Updated code attached:
```python3
import logging

import redis

from llama_index import SimpleDirectoryReader, VectorStoreIndex
from llama_index.storage.docstore import RedisDocumentStore
from llama_index.storage.index_store import RedisIndexStore
from llama_index.storage.storage_context import StorageContext
from llama_index.vector_stores import RedisVectorStore

logging.basicConfig(level=logging.INFO)

redis_host = "localhost"
redis_port = 6379
namespace = "llama_index"


def metadata(path):
    return {"path": path}


# Clean up the database before overwriting
# Connect to Redis
redis_client = redis.Redis(host=redis_host, port=redis_port)

# Delete all keys in the namespace
redis_client.delete(f"{namespace}/doc")
redis_client.delete(f"{namespace}/index")
redis_client.delete(f"{namespace}/metadata")
redis_client.close()

documents = SimpleDirectoryReader(
    "data/paul_graham/", file_metadata=metadata
).load_data()

docstore = RedisDocumentStore.from_host_and_port(
    host=redis_host, port=redis_port, namespace=namespace
)
docstore.add_documents(documents)
index_store = RedisIndexStore.from_host_and_port(
    host=redis_host, port=redis_port, namespace=namespace
)
vector_store = RedisVectorStore(
    index_name=namespace,
    index_prefix=namespace,
    redis_url=f"redis://{redis_host}:{redis_port}",
    overwrite=True,
)
storage_context = StorageContext.from_defaults(
    vector_store=vector_store, index_store=index_store, docstore=docstore
)
index = VectorStoreIndex.from_documents(
    documents, storage_context=storage_context, store_nodes_override=True
)

query_engine = index.as_query_engine()
response = query_engine.query("What did the author learn?")
print(response, type(response))
for n in response.source_nodes:
    print(n.node.metadata)
```

logan-markewich commented on llama_index/indices/vector_store/retrievers/retriever.py:
```diff
@@ -112,6 +112,16 @@ def _retrieve(
+                    # Fix metadata for nodes that have source node in the docstore
+                    elif (
+                        source_node is not None
+                        and source_node.node_id in self._docstore.docs
```
`self._docstore.docs` is a pretty expensive operation. Can we call this once and reuse it?

Also, I have no idea how the code gets here when using milvus -- the top level if statement will return false

HollowMan6 replied:
Done! The code applies to both Milvus and Redis vector store since they won't show metadata anyway, so I'm assuming that this fix applies to all the vector store database except the Simple Vector Store.

logan-markewich commented:
Yea milvus has some weird stuff going on -- I'm still confused with how `store_nodes_override` is not returning the metadata, and I want to take the time to debug that before merging this

logan-markewich commented:
@HollowMan6 tangentially related https://github.com/jerryjliu/llama_index/pull/6667

HollowMan6 commented:
@logan-markewich  Thanks. Now the Milvus returns metadata when `store_nodes_override=False`, but still, the response won't get printed when `store_nodes_override=True`, showing `None`.

However, I would mention that Redis still doesn't return the metadata no matter `store_nodes_override`, and I'm not sure whether it's also the case for other vector stores listed here: https://gpt-index.readthedocs.io/en/latest/how_to/integrations/vector_stores.html#using-a-vector-store-as-an-index

logan-markewich commented:
@HollowMan6 the reason for redis vector store returning none is because it doesn't store metadata (there is a todo comment to add in the code lol)

For the general problem with `store_nodes_override` not working, I discovered that there is a few bugs due to badly named variables. `query.doc_ids` refers to `node_ids`, but some vector store implement this improperly 

I will fix this in a separate PR 👍🏻

HollowMan6 commented:
@logan-markewich  I see, but if we need a separate implementation for vector store to store metadata, why not just use the metadata in the docstore, just like this PR instead? This PR can act as a fail-safe for the vector store metadata implementation just like the Milvus case.

> the reason for redis vector store returning none is because it doesn't store metadata (there is a todo comment to add in the code lol)

logan-markewich commented:
@HollowMan6 because with most vector stores, the docstore and and index store are not actually used at all.

See this line of code when building the index: https://github.com/jerryjliu/llama_index/blob/8261f51773b97c9c023c89ef33b2d4245a6b5228/llama_index/indices/vector_store/base.py#L167

The key part of this is that then users don't have to worry about setting up a docstore or index store, which tbh is a pretty convenient design in most cases.

With the recent node refactors, it is much easier to store the node data in the vector db since you can serialize it entirely. It could be easily added to redis tbh, it just got missed in the refactor.

The thing about this PR is it still feels a little hacky to me, as it ignores other design issues that should be addressed instead 🤔  Which is why I spent so much time debugging the real issue here lol

HollowMan6 commented:
Okay, thanks! Let me know once the Redis metadata is ready, then I can close this PR (I currently rely on the metadata and this is a key feature used in our project).

logan-markewich commented:
@HollowMan6 the last few PRs I merged should address your concerns. Let me know if anything else comes up 💪🏻
